crates/sui-keys/src/external.rs

```diff
@@ -0,0 +1,944 @@
+        })?;
+        let msg = general_purpose::STANDARD.encode(&msg);
+
+        let sign_request = SignRequest {
+            key_id: key_id.clone(),
+            msg,
+            intent: Some(intent),
+        };
+
+        let result = self
+            .exec(
+                &signer,
+                "sign",
+                serde_json::to_value(sign_request).map_err(|e| {
+                    signature::Error::from_source(anyhow!(
+                        "Failed to serialize sign request: {}",
+                        e
+                    ))
+                })?,
+            )
+            .await
+            .map_err(|e| signature::Error::from_source(anyhow!("Failed to sign message: {}", e)))?;
+
+        let result: SignResponse = serde_json::from_value(result).map_err(|e| {
+            signature::Error::from_source(anyhow!("Failed to parse sign response: {}", e))
+        })?;
+
+        let signature = Signature::decode_base64(&result.signature).unwrap();
+        Ok(signature)
+    }
+
+    fn addresses_with_alias(&self) -> Vec<(&SuiAddress, &Alias)> {
+        let mut addresses = Vec::new();
+        for (address, alias) in &self.aliases {
+            addresses.push((address, alias));
+        }
+        addresses
+    }
+
+    fn aliases(&self) -> Vec<&Alias> {
+        let mut aliases = Vec::new();
+        for alias in self.aliases.values() {
+            aliases.push(alias);
+        }
+        aliases
+    }
+
+    fn aliases_mut(&mut self) -> Vec<&mut Alias> {
+        let mut aliases = Vec::new();
+        for alias in self.aliases.values_mut() {
+            aliases.push(alias);
+        }
+        aliases
+    }
+
+    fn create_alias(&self, alias: Option<String>) -> Result<String, Error> {
+        match alias {
+            Some(a) if self.alias_exists(&a) => {
+                bail!("Alias {a} already exists. Please choose another alias.")
+            }
+            Some(a) => validate_alias(&a),
+            None => Ok(random_name(
+                &self
+                    .aliases()
+                    .into_iter()
+                    .map(|x| x.alias.to_string())
+                    .collect::<HashSet<_>>(),
+            )),
+        }
+    }
+
+    async fn update_alias(
+        &mut self,
+        old_alias: &str,
+        new_alias: Option<&str>,
+    ) -> Result<String, Error> {
+        if !self.alias_exists(old_alias) {
+            bail!("The provided alias {old_alias} does not exist");
+        }
+        let new_alias_name = self.create_alias(new_alias.map(str::to_string))?;
+        for a in self.aliases_mut() {
+            if a.alias == old_alias {
+                let pk = &a.public_key_base64;
+                *a = Alias {
+                    alias: new_alias_name.clone(),
+                    public_key_base64: pk.clone(),
+                };
+            }
+        }
+        Ok(new_alias_name)
+    }
+}
```

Comment:
are these alias impl already exists for other keystores? i wonder if theres a way not to duplicate them but implement at the trait level.